### PR TITLE
Add OCI building and publishing to golang devtools

### PR DESCRIPTION
**Pull request by aucampia**
Add tools and targets for building and publishing OCI images to the
golang devtools, more details can be seen in the README.

### CODE REVIEW

**Comment by nikitabarskov** [comment]
[command.log](https://github.com/coopnorge/engineering-docker-images/files/9269433/command.log)

**Comment by AtzeDeVries** [comment]
@nikitabarskov the command.log is empty. I assuming these are the commands you used to test this pr. Can you reshare this so i can mimic testing you did on a M1?

**Comment by aucampia** [comment]
@AtzeDeVries to test this on your computer you should be able to just follow the instructions in [images/devtools-golang-v1beta1/context/README.md](https://github.com/coopnorge/engineering-docker-images/blob/iwana-20220804T2338-golang_oci/images/devtools-golang-v1beta1/context/README.md)

But in brief:

```bash
cd ./images/devtools-golang-v1beta1/context/
docker compose build
docker compose run --rm devtools
```

**Review by stephendrake — APPROVED** [review]

**Comment by nikitabarskov** [comment]
> @nikitabarskov the command.log is empty. I assuming these are the commands you used to test this pr. Can you reshare this so i can mimic testing you did on a M1?
@AtzeDeVries 
https://gist.github.com/nikitabarskov/7e3bb2892b68fe94b50178b0449f6b0f

**Comment by nikitabarskov** [comment]
@aucampia I ran this, works fine (Intel Mac).

> @AtzeDeVries to test this on your computer you should be able to just follow the instructions in [images/devtools-golang-v1beta1/context/README.md](https://github.com/coopnorge/engineering-docker-images/blob/iwana-20220804T2338-golang_oci/images/devtools-golang-v1beta1/context/README.md)
> 
> But in brief:
> 
> ```shell
> cd ./images/devtools-golang-v1beta1/context/
> docker compose build
> docker compose run --rm devtools
> ```

File: images/devtools-golang-v1beta1/context/Dockerfile [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
 FROM docker.io/mikefarah/yq:4@sha256:856f7ab12c58608d422a742a4917e6998ad065d9d48a5b1cd69c3ce8fa80f3fa AS yq
-
-
-# golangci-lint does not work with go1.18
-# https://github.com/golangci/golangci-lint/issues/2649
+FROM docker.io/hadolint/hadolint:v2.10.0@sha256:93f0afd12c3be5d732227c0226dd8e7bb84f79319a773d7f8519e55f958ba415 AS hadolint
+FROM docker.io/goodwithtech/dockle:v0.4.6@sha256:534ecfe2204403e1b563d61b255ddd4762b20c9d72ce7af239fb1135be6c9569 AS dockle
+FROM docker.io/moby/buildkit:v0.10.3-rootless@sha256:22e39442689ddbfd938892b33a12cc9bc76c7e71e07cbaa20c8272a7aab2fd59 AS buildkit
+FROM gcr.io/go-containerregistry/crane:v0.11.0@sha256:40823185f184cfc2fed19318988ad5119f4d2fd1eadee51d8b2b365016a6c03b AS crane
 FROM golang:1.18.5-bullseye@sha256:ac2ea4e233e5f1bb4407893f58d4e33d908945d4163e5848160ab588d1d6a456 AS base
```
Comment (nikitabarskov) [inline images/devtools-golang-v1beta1/context/Dockerfile:8]: @aucampia fully qualified docker image reference?

https://playbook.internal.coop/guidelines/security/docker-images-and-containers.html?h=docker#use-fully-qualified-docker-image-references

File: images/devtools-golang-v1beta1/context/README.md [hunk 2]
```diff
@@ -1,4 +1,31 @@
...
+cd ./images/devtools-golang-v1beta1/context/
+
+# build the image
 docker compose build
-docker compose run devtools bash
+
+# run the default target
+docker compose run --rm devtools
+
+
```
Comment (nikitabarskov) [inline images/devtools-golang-v1beta1/context/README.md:10]: Too many blank lines in this file :)

**Review by nikitabarskov — CHANGES_REQUESTED** [review]

File: images/devtools-golang-v1beta1/context/README.md (hunk 2 above)
Comment (aucampia) [reply images/devtools-golang-v1beta1/context/README.md:10]: Fair enough, will fix.

Comment (aucampia) [reply images/devtools-golang-v1beta1/context/README.md]: Cleaned it up now.

File: images/devtools-golang-v1beta1/context/Dockerfile (hunk 1 above)
Comment (aucampia) [reply images/devtools-golang-v1beta1/context/Dockerfile]: fixed

**Review by stephendrake — DISMISSED** [review]

**Comment by aucampia** [comment]
@nikitabarskov still pending on your review which was marked as requesting changes.

**Comment by AtzeDeVries** [comment]
For me this fails,
```
➜ cd ./images/devtools-golang-v1beta1/context/
  docker compose build
  docker compose run --rm devtools
[+] Building 4.7s (10/10) FINISHED
 => [internal] load build definition from Dockerfile                                                               0.0s
 => => transferring dockerfile: 3.82kB                                                                             0.0s
 => [internal] load .dockerignore                                                                                  0.0s
 => => transferring context: 2B                                                                                    0.0s
 => [internal] load metadata for docker.io/library/golang:1.18.5-bullseye@sha256:ac2ea4e233e5f1bb4407893f58d4e33d  4.7s
 => [internal] load metadata for docker.io/hadolint/hadolint:v2.10.0@sha256:93f0afd12c3be5d732227c0226dd8e7bb84f7  4.3s
 => CANCELED [internal] load metadata for docker.io/moby/buildkit:v0.10.3-rootless@sha256:22e39442689ddbfd938892b  4.6s
 => ERROR [internal] load metadata for docker.io/goodwithtech/dockle:v0.4.6@sha256:534ecfe2204403e1b563d61b255ddd  4.6s
 => [internal] load metadata for docker.io/golangci/golangci-lint:v1.48.0-alpine@sha256:e9ce901b7974d641d6288b070  4.7s
 => [internal] load metadata for gcr.io/go-containerregistry/crane:v0.11.0@sha256:40823185f184cfc2fed19318988ad51  3.1s
 => [internal] load metadata for docker.io/safewaters/docker-lock:latest@sha256:432d90ddc2891f4845241adc63e5eef2d  4.1s
 => CANCELED [internal] load metadata for docker.io/mikefarah/yq:4@sha256:856f7ab12c58608d422a742a4917e6998ad065d  4.6s
------
 > [internal] load metadata for docker.io/goodwithtech/dockle:v0.4.6@sha256:534ecfe2204403e1b563d61b255ddd4762b20c9d72ce7af239fb1135be6c9569:
------
failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition: no match for platform in manifest sha256:534ecfe2204403e1b563d61b255ddd4762b20c9d72ce7af239fb1135be6c9569: not found
[+] Running 2/0
 ⠿ Network context_default          Created                                                                        0.0s
 ⠿ Volume "context_xdg-cache-home"  Created                                                                        0.0s
[+] Building 0.1s (10/10) FINISHED
 => [internal] load build definition from Dockerfile                                                               0.0s
 => => transferring dockerfile: 32B                                                                                0.0s
 => [internal] load .dockerignore                                                                                  0.0s
 => => transferring context: 2B                                                                                    0.0s
 => [internal] load metadata for docker.io/library/golang:1.18.5-bullseye@sha256:ac2ea4e233e5f1bb4407893f58d4e33d  0.0s
 => [internal] load metadata for docker.io/hadolint/hadolint:v2.10.0@sha256:93f0afd12c3be5d732227c0226dd8e7bb84f7  0.0s
 => ERROR [internal] load metadata for docker.io/goodwithtech/dockle:v0.4.6@sha256:534ecfe2204403e1b563d61b255ddd  0.0s
 => [internal] load metadata for docker.io/safewaters/docker-lock:latest@sha256:432d90ddc2891f4845241adc63e5eef2d  0.0s
 => [internal] load metadata for gcr.io/go-containerregistry/crane:v0.11.0@sha256:40823185f184cfc2fed19318988ad51  0.0s
 => [internal] load metadata for docker.io/golangci/golangci-lint:v1.48.0-alpine@sha256:e9ce901b7974d641d6288b070  0.0s
 => CANCELED [internal] load metadata for docker.io/moby/buildkit:v0.10.3-rootless@sha256:22e39442689ddbfd938892b  0.0s
 => CANCELED [internal] load metadata for docker.io/mikefarah/yq:4@sha256:856f7ab12c58608d422a742a4917e6998ad065d  0.1s
------
 > [internal] load metadata for docker.io/goodwithtech/dockle:v0.4.6@sha256:534ecfe2204403e1b563d61b255ddd4762b20c9d72ce7af239fb1135be6c9569:
------
failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition: no match for platform in manifest sha256:534ecfe2204403e1b563d61b255ddd4762b20c9d72ce7af239fb1135be6c9569: not found
```

I guess this has to do with dockle

**Comment by AtzeDeVries** [comment]
fixed build, but output is not fully the same as Nikita. It does not seem to run the test, but i did not research yet.  https://gist.github.com/AtzeDeVries/69d7d87f2b96e900b3cb16c3b33395ad

**Comment by aucampia** [comment]
> fixed build, but output is not fully the same as Nikita. It does not seem to run the test, but i did not research yet. https://gist.github.com/AtzeDeVries/69d7d87f2b96e900b3cb16c3b33395ad

How did you fix the build?

Nikita's run was on an older version I think, but also he was running the OCI build which is:


```bash
# change into the image context directory
cd ./images/devtools-golang-v1beta1/context/

# build and validate the OCI image
docker compose run --rm -e APP_DOCKERFILE=build/package/Dockerfile.example devtools validate VERBOSE=all; \
cat var/oci_images/stage-runtime.oci.tar | docker image load
docker run --rm -it ocreg.invalid/prototype:latest 1 2 3
docker image save ocreg.invalid/prototype:latest | dlayer
```

Again from same readme linked earlier.

**Comment by aucampia** [comment]
> > fixed build, but output is not fully the same as Nikita. It does not seem to run the test, but i did not research yet. https://gist.github.com/AtzeDeVries/69d7d87f2b96e900b3cb16c3b33395ad
> 
> How did you fix the build?

Okay nevermind, I saw your commit now, I also fixed hadolint's complaint

**Comment by aucampia** [comment]
Also to run the tests (from [`images/devtools-golang-v1beta1/tests/README.md`](https://github.com/coopnorge/engineering-docker-images/blob/iwana-20220804T2338-golang_oci/images/devtools-golang-v1beta1/tests/README.md))

```bash
## Run commands from repo root

# Build image.
make IMAGE_NAMES=devtools-golang-v1beta1 build

# Test image.
IMAGE_UNDER_TEST=ocreg.invalid/coopnorge/engineering/image/devtools-golang-v1beta1:built \
poetry run pytest images/devtools-golang-v1beta1/tests
```

this requires poetry though 

https://github.com/coopnorge/engineering-docker-images/blob/99382479a0a6ad08eada5a1a7050baaa65bdf739/README.md?plain=1#L147-L150

**Review by stephendrake — APPROVED** [review]

**Pull request merged**